File: src/main/java/ro/sopa/statistifier/importer/ListeningHistoryJob.java
package ro.sopa.statistifier.importer;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.persistence.EntityManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import ro.sopa.statistifier.UserImportException;
import ro.sopa.statistifier.api.client.LastFMClient;
import ro.sopa.statistifier.api.model.ListeningHistory;
import ro.sopa.statistifier.api.model.Track;
import ro.sopa.statistifier.db.model.TrackListen;
import ro.sopa.statistifier.db.repository.TrackListenRepository;
import ro.sopa.statistifier.job.JobDescriptor;
import ro.sopa.statistifier.job.JobRegistry;
import ro.sopa.statistifier.job.JobType;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

@Component
public class ListeningHistoryJob {

    private static final Logger logger = LoggerFactory.getLogger(ListeningHistoryJob.class);

    @Autowired
    LastFMClient lastFMClient;

    @Value("${lastfm.api.key}")
    private String apiKey;

    @Autowired
    TrackListenRepository trackListenRepository;

    @Autowired
    EntityManager entityManager;

    @Autowired
    JobRegistry jobRegistry;

    ObjectMapper objectMapper = new ObjectMapper();

    public void downloadAndPersistHistory(String username, Integer startPage, JobDescriptor jobDescriptor) {

        ZonedDateTime dateOfLastListen = trackListenRepository.findLatestListenDate(username);

        boolean shouldContinue = true;

        jobRegistry.setStep(jobDescriptor, startPage);

        logger.info("Fetching page " + startPage + "...");
        ListeningHistory page = lastFMClient.callRecentTracks(apiKey, username, startPage);
        shouldContinue = persist(page, username, dateOfLastListen);

        String nrPagesStr = page.getRecenttracks().getAttr().getTotalPages();
        Integer nrPages = Integer.parseInt(nrPagesStr);

        jobRegistry.setTotalStep(jobDescriptor, nrPages);

        // crashed @ 414
        //  Data truncation: Incorrect datetime value: '2015-03-29T03:56:00' for column 'date' at row 1
        for(int i = startPage + 1; i <= nrPages; i++) {
            if(!shouldContinue) {
                logger.info("Stopped because scrobbles are up to date.");
                break;
            }

            jobRegistry.setStep(jobDescriptor, i);

            logger.info("Fetching page " + i + " of " + nrPages + "...");
            page = lastFMClient.callRecentTracks(apiKey, username, i);
            shouldContinue = persist(page, username, dateOfLastListen);
            entityManager.clear();
        }

        logger.info("Finished!");
    }

    private boolean persist(ListeningHistory page, String username, ZonedDateTime dateOfLastListen) {
        for(int i = 0; i < page.getRecenttracks().getTrack().size(); i++) {
            Track t = page.getRecenttracks().getTrack().get(i);

            try {
                boolean shouldContinue = persist(t, username, dateOfLastListen);
                if(!shouldContinue) {
                    return false;
                }
            } catch(Throwable thr) {
                logger.error("Error while persisting track on page " + page.getRecenttracks().getAttr().getPage() + " with index = " + i);
                try {
                    logger.error(objectMapper.writeValueAsString(t));
                } catch (JsonProcessingException e) {
                    throw new UserImportException("Error while persisting listen history", e);
                }
            }
        }

        return true;
    }

    private boolean persist(Track track, String username, ZonedDateTime dateOfLastListen) {

        TrackListen listen = mapTrack(track, username);

        if(listen.getDate().equals(dateOfLastListen)) {
            return false;
        }

        trackListenRepository.save(listen);

        return true;
    }

    private TrackListen mapTrack(Track track, String username) {
        TrackListen listen = new TrackListen();

        listen.setUsername(username);
        listen.setAlbum(track.getAlbum().getText());
        listen.setAlbumId(track.getAlbum().getMbid());
        listen.setArtist(track.getArtist().getText());
        listen.setArtistId(track.getArtist().getMbid());
        listen.setTrack(track.getName());
        listen.setDate(mapDate(track.getDate().getText()));

        return listen;
    }

    private static ZonedDateTime mapDate(String date) {
        return ZonedDateTime.parse(date, DateTimeFormatter.ofPattern("dd MMM yyyy, HH:mm").withZone(ZoneId.of("UTC")).withLocale(Locale.US));
    }

}

File: src/main/kotlin/ro/sopa/lastfm/importer/ListeningHistoryJob.kt
package ro.sopa.lastfm.importer

import com.fasterxml.jackson.core.JsonProcessingException
import com.google.gson.Gson
import jakarta.persistence.EntityManager
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import ro.sopa.lastfm.api.client.LastFMClient
import ro.sopa.lastfm.api.model.ListeningHistory
import ro.sopa.lastfm.api.model.Track
import ro.sopa.lastfm.db.model.Artist
import ro.sopa.lastfm.db.model.ArtistAlias
import ro.sopa.lastfm.db.model.TrackListen
import ro.sopa.lastfm.db.model.User
import ro.sopa.lastfm.db.repository.ArtistAliasRepository
import ro.sopa.lastfm.db.repository.ArtistRepository
import ro.sopa.lastfm.db.repository.TrackListenRepository
import ro.sopa.lastfm.db.repository.UserRepository
import ro.sopa.lastfm.exception.UserImportException
import ro.sopa.lastfm.job.JobDescriptor
import ro.sopa.lastfm.job.JobRegistry
import ro.sopa.lastfm.service.ArtistService
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.*

@Component
class ListeningHistoryJob(val lastFMClient: LastFMClient,
                          val trackListenRepository: TrackListenRepository,
                          val artistRepository: ArtistRepository,
                          val artistAliasRepository: ArtistAliasRepository,
                          val entityManager: EntityManager,
                          val artistService: ArtistService,
                          val userRepository: UserRepository,
                          val jobRegistry: JobRegistry
) {

    private val logger = LoggerFactory.getLogger(ListeningHistoryJob::class.java)

    @Value("\${lastfm.api.key}")
    private lateinit var apiKey: String

    private val gson = Gson();

    fun downloadAndPersistHistory(username: String, startPage: Int, jobDescriptor: JobDescriptor) {

        var maybeUser = userRepository.findByUsername(username)
        val user = maybeUser ?: createUser(username)

        val dateOfLastListen: ZonedDateTime? = trackListenRepository.findLatestListenDate(user.id!!)
        var shouldContinue = true

        val userId = getOrCreateUser(username)

        jobRegistry.setStep(jobDescriptor, startPage)

        logger.info("Fetching page $startPage...")

        var page: ListeningHistory = lastFMClient.callRecentTracks(apiKey, username, startPage)
        shouldContinue = persist(page, userId, dateOfLastListen)
        val nrPagesStr: String? = page.recenttracks?.attr?.totalPages

        val nrPages = nrPagesStr?.toInt() ?: 1
        jobRegistry.setTotalStep(jobDescriptor, nrPages)

        for (i in startPage + 1..nrPages) {
            if (!shouldContinue) {
                logger.info("Stopped because scrobbles are up to date.")
                break
            }
            jobRegistry.setStep(jobDescriptor, i)
            logger.info("Fetching page $i of $nrPages...")
            page = lastFMClient.callRecentTracks(apiKey, username, i)
            shouldContinue = persist(page, userId, dateOfLastListen)
            entityManager.clear()
        }
        logger.info("Finished!")
    }

    private fun getOrCreateUser(username: String): Int {
        val user = userRepository.findByUsername(username);
        return user?.id ?: createUser(username).id!!;
    }

    private fun createUser(username: String): User {
        val user = User(null, username)
        return userRepository.save(user)
    }


    private fun persist(page: ListeningHistory, userId: Int, dateOfLastListen: ZonedDateTime?): Boolean {
        for (i in 0..page.recenttracks?.track?.size!! - 1) {
            val t: Track = page.recenttracks?.track?.get(i) ?: throw RuntimeException("No tracks returned");
            try {
                val shouldContinue = persist(t, userId, dateOfLastListen)
                if (!shouldContinue) {
                    return false
                }
            } catch (thr: Throwable) {
                logger.error(
                    "Error while persisting track on page " + page.recenttracks?.attr?.page + " with index = " + i, thr
                )
                try {
                    logger.error(gson.toJson(t))
                } catch (e: JsonProcessingException) {
                    throw UserImportException("Error while persisting listen history", e)
                }
            }
        }
        return true
    }

    private fun persist(track: Track, userId: Int, dateOfLastListen: ZonedDateTime?): Boolean {
        val listen: TrackListen = mapTrack(track, userId)
        if (listen.date == dateOfLastListen) {
            return false
        }
        trackListenRepository.save(listen)
        return true
    }

    private fun mapTrack(track: Track, userId: Int): TrackListen {

        val artistId = findArtist(track);

        val listen = TrackListen(track = track.name ?: "Unknown", userId = userId, artistId = artistId)
        listen.album = track.album?.text
        listen.albumId = track.album?.mbid
        listen.date = mapDate(track.date?.text!!)

        return listen
    }

    private fun findArtist(track: Track): Int {
        if(track.artist != null) {
            var artistId = artistAliasRepository.findArtistIdByAlias(track.artist?.text!!)

            return artistId ?: handleMissingArtistAlias(track);
        }

        throw RuntimeException("Invalid artist received")
    }

    private fun handleMissingArtistAlias(track: Track): Int {
        val corrections = lastFMClient.callCorrections(apiKey, track.artist?.text!!)

        val correctName = corrections.corrections?.correction?.artist?.name!!

        var artist = artistRepository.findByName(correctName) ?: handleMissingArtist(corrections.corrections?.correction?.artist!!)

        return createArtistAlias(artist, track.artist?.text!!);
    }

    private fun handleMissingArtist(artist: ro.sopa.lastfm.api.model.correction.Artist): Artist {
        val artistDb = Artist(null, artist.name!!, artist.mbid)
        val savedArtist = artistRepository.save(artistDb)

        artistService.downloadTags(savedArtist);

        return savedArtist;
    }

    private fun createArtistAlias(artist: Artist, aliasName: String): Int {
        val alias = ArtistAlias(null, artist.id!!, aliasName)
        val persistedAlias = artistAliasRepository.save(alias)
        return persistedAlias.id!!.toInt()
    }

    private fun mapDate(date: String): ZonedDateTime? = ZonedDateTime.parse(date,
            DateTimeFormatter.ofPattern("dd MMM yyyy, HH:mm").withZone(ZoneId.of("UTC")).withLocale(Locale.US))
}